src/libraries/System.Diagnostics.EventLog/src/Messages/readme.md

```diff
@@ -0,0 +1,15 @@
+These files are used to produce an Event Message File.
+
+For more information see https://docs.microsoft.com/en-us/windows/win32/eventlog/message-files.
+
+The design of the EventLog class is to allow for the registration of event sources without specifying message files.
+
+In the case an event source does not specify it's own message file, EventLog just provides a default message file 
+with 64K message IDs all that just pass through the first insertion string.  This allow the event source to still
+use IDs for messages, but doesn't require the caller to actually pass a message file in order to achieve this.
+
+The process for producing the message file requires mc.exe and rc.exe which do not work cross-platform, and they 
```

Comment:
If cross-plat is a concern, would a checked in binary blob work for source build?

FWIW, the [RES file format](https://docs.microsoft.com/en-us/windows/win32/menurc/resource-file-formats) is simple and we have a [C#-based emitter](https://github.com/dotnet/runtimelab/blob/dab84ea91b9c03492c0e77b2eef62e9997562213/src/coreclr/tools/aot/ILCompiler.Build.Tasks/DumpNativeResources.cs) for it.

The MSG00001.bin blob embedded in the RES file looks pretty regular and shouldn't be difficult to reverse engineer and generate either (I couldn't quickly find official docs).

Reply:
That's a cool tool.  I was able to find the docs on that binary format: https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-message_resource_data

So for our sequence of contiguous message IDs it would look like
```
MESSAGE_RESOURCE_DATA 
{ 
  NumberOfBlocks = 1,
  Blocks = 
  [
    MESSAGE_RESOURCE_BLOCK
    {
      LowId = 0,
      HighId = 65535, // 64K messages
      OffsetToEntries = 16 // address immediately following this structure
    }
    [
      MESSAGE_RESOURCE_ENTRY  // 64K of these
      {
        Length = 12,  // length of this structure in bytes padded to DWORD boundary
        Flags = 1,  // Unicode
        Text = '%', '1', '\0', '\0' (padding)
      },
      ...
    ]
 ]
```

We could use ANSI encoding instead (Flags = 0), and save 4 bytes per entry, or 256 KB in total DLL size.

Interestingly I found that mc.exe doesn't emit ansi, even when you tell it `-c ansi`, but you can use the deprecated `-A` switch to get it to do so.  Perhaps that's a bug?  Makes me reluctant to change the format here.

So this seems feasible, but is it worth it?  We'd have to maintain a task for generating this resource, a resource that hasn't changed in many years, in a technology we consider legacy.  I wonder if we could instead think of this like an image: it's a checked in *content* file.  We just happen to have a way to auto-generate it.  Thoughts @jkotas @danmosemsft?

Reply:
Also @dseefeld, since this is a source-build question.

Effectively the question is:
Can we consider a checked in `.res` binary file source build friendly, or should we write a custom tool to generate it?

Reply:
Do we have precedent for checking in binary files (even with generator code to match)? I admit it is unappealing to do it in the build, but I have also seen such code rot through lack of use.

Reply:
I did a scan of the repo for binary files and found quite a few.  Here are some notable cases that match this pattern for precedent, some are hiding binary data inside C# files but its a similar thing:
https://github.com/dotnet/runtime/blob/0bddd531f4deed535ad89b0ffc9c7607a7d37359/src/libraries/System.Private.Xml/src/System/Xml/XmlCharType.cs#L199
https://github.com/dotnet/runtime/tree/master/src/libraries/System.Drawing.Common/src/Resources/System/Drawing
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Text.Encoding.CodePages/src/Data/codepages.nlp
https://github.com/dotnet/runtime/blob/e1297870e5b900405a3367da3db62bb9a6934972/src/libraries/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfoData.cs#L10

This resource was handled in the same way in .NETFramework and was never changed since at least 2002, maybe never changed.  I found a very nice internal document written by @KrzysztofCwalina the same year that describes this feature as a "trick" 🤔 
 
Now I don't mind writing a generator task for this, quite frankly I quite enjoyed researching the format and could probably whip up a generator for it, but I don't think its really worth it from a cost/benefit perspective.

Reply:
@ericstj, is the `.res` file included in Windows only?

Reply:
It's only included in the Windows assembly, correct.  This is a package so it contains both windows and non-windows bits.  When consumed on non-windows it will not use this resource.

Reply:
This shouldn't be an issue for source-build.